src/librustc_mir/transform/const_prop.rs

```diff
@@ -459,97 +460,81 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
     ) -> Option<Const<'tcx>> {
         let span = source_info.span;
 
-        // if this isn't a supported operation, then return None
-        match rvalue {
-            Rvalue::Repeat(..) |
-            Rvalue::Aggregate(..) |
-            Rvalue::NullaryOp(NullOp::Box, _) |
-            Rvalue::Discriminant(..) => return None,
-
-            Rvalue::Use(_) |
-            Rvalue::Len(_) |
-            Rvalue::Cast(..) |
-            Rvalue::NullaryOp(..) |
-            Rvalue::CheckedBinaryOp(..) |
-            Rvalue::Ref(..) |
-            Rvalue::UnaryOp(..) |
-            Rvalue::BinaryOp(..) => { }
-        }
-
         // perform any special checking for specific Rvalue types
```

Comment:
It took me quite some time to decipher this code... I think this could have better comments.

I think what is going on is that we end up calling into the Miri engine to interpret this line, and in an ideal world we might just do that. But instead we first do a bunch of checks to (a) provide diagnostics, and (b) hack around ICEs in the Miri engine. However, I still don't understand why some branches early-abort (`return None`) while others don't.